README.md

```diff
@@ -999,74 +894,141 @@ also avoids eager string encoding onto the stack and the need for NUL
 termination, allowing string contents to be written to memory exactly
 where they are needed.
 
+### Why should WebAssembly strings be able to represent isolated surrogates?
+
+The main motivation is to support source languages with WTF-16 strings
+(e.g. Java, Kotlin, C#).  JVM-based and CLR-based languages treat
+strings as sequences of 16-bit code units.  Sometimes programs written
+in e.g. Java will decode these sequences into codepoints encoded as
+WTF-16, but not always.  Many common algorithms can be performed
+directly on the code units, for example prefix matching.  Therefore to
+efficiently support Java and friends when compiled to WebAssembly, we
+need to support this view of strings as sequences of any 16-bit code
+units, without any validity constraints that enforce that surrogates
+always be properly paired.  This is the main reason to support WTF-16
+rather than just UTF-16.
+
+An important secondary reason is interoperability with JavaScript hosts.
+For zero-copy interoperation with JavaScript and DOM facilities, it
+would be good if for `stringref` to have the same semantics as a
+JavaScript string, which like Java is an arbitrary sequence of 16-bit
+code units.
+
+Isolated surrogates are rare in JavaScript, but can occur via:
+  - Reading invalid UTF-16 from external sources.  However this is not
+    common, as isolated surrogates are not valid UTF-16.
```

Comment:
Is this saying "as invalid UTF-16 is not valid UTF-16"? Not sure how to phrase better